mipengine/controller/node_registry.py

```diff
@@ -7,28 +7,35 @@
 
 
 class NodeRegistry(BaseModel):
-    nodes: Dict[str, NodeInfo] = {}
+    def __init__(self, nodes_info=None):
+        if nodes_info:
+            global_node = [
+                node_info
+                for node_info in nodes_info
+                if node_info.role == NodeRole.GLOBALNODE
+            ][0]
+            local_nodes = [
+                node_info
+                for node_info in nodes_info
+                if node_info.role == NodeRole.LOCALNODE
+            ]
+            super().__init__(global_node=global_node, local_nodes=local_nodes)
+        else:
+            super().__init__()
+
+    global_node: NodeInfo = None
+    local_nodes: List[NodeInfo] = []
 
     class Config:
         allow_mutation = False
         arbitrary_types_allowed = True
 
-    def _get_all_global_nodes(self) -> Dict[str, NodeInfo]:
-        return {
-            node_id: node_info
-            for node_id, node_info in self.nodes.items()
-            if node_info.role == NodeRole.GLOBALNODE
-        }
-
-    def get_global_node(self) -> NodeInfo:
-        return list(self._get_all_global_nodes().values())[0]
-
-    def get_all_local_nodes(self) -> Dict[str, NodeInfo]:
-        return {
-            node_id: node_info
-            for node_id, node_info in self.nodes.items()
-            if node_info.role == NodeRole.LOCALNODE
-        }
+    def get_nodes(self) -> List[NodeInfo]:
+        return (
+            self.local_nodes + [self.global_node]
+            if self.global_node and self.local_nodes
```

Comment:
this will return empty even if there is a global node but no local nodes or if there are local nodes but no global node